daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/ContractStateMachineSpec.scala

```diff
@@ -0,0 +1,651 @@
+      s"pass $name" when {
+        expected.foreach { case (mode, expectedResult) =>
+          s"mode $mode" in {
+            // We use `Unit` instead of `NodeId` so that we don't have to fiddle with node ids
+            val ksm = new ContractStateMachine[Unit](mode)
+            val actualResolver: KeyResolver =
+              if (mode == ContractKeyUniquenessMode.Strict) Map.empty else resolver
+            val result = visitSubtrees(ksm)(tx.nodes, tx.roots.toSeq, actualResolver, ksm.initial)
+
+            (result, expectedResult) match {
+              case (Left(err1), Left(err2)) => err1 shouldBe err2
+              case (Right(state), Right((gkI, activeState))) =>
+                withClue("global key inputs") {
+                  state.globalKeyInputs shouldBe gkI
+                }
+                withClue("active state") {
+                  state.activeState shouldBe activeState
+                }
+                state.rollbackStack shouldBe List.empty
+              case _ => fail(s"$result was not equal to $expectedResult")
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private def children(node: Node): ImmArray[NodeId] = node match {
+    case _: Node.Create | _: Node.Fetch | _: Node.LookupByKey => ImmArray.empty[NodeId]
+    case exercise: Node.Exercise => exercise.children
+    case rollback: Node.Rollback => rollback.children
+  }
+
+  private def visitSubtree(ksm: ContractStateMachine[Unit])(
+      nodes: Map[NodeId, Node],
+      root: NodeId,
+      resolver: KeyResolver,
+      state: ksm.State,
+  ): Either[Transaction.KeyInputError, ksm.State] = {
+    val node = nodes(root)
+    for {
+      next <- node match {
+        case create: Node.Create =>
+          state.visitCreate(create.templateId, create.coid, create.key)
+        case exercise: Node.Exercise => state.handleExercise((), exercise)
+        case fetch: Node.Fetch => state.handleFetch(fetch)
+        case lookup: Node.LookupByKey =>
+          val gkey = GlobalKey.assertBuild(lookup.templateId, lookup.key.key)
+          if (state.mode == ContractKeyUniquenessMode.Strict)
+            state.handleLookup(lookup)
+          else
+            state.handleLookupWith(lookup, resolver(gkey))
+        case _: Node.Rollback =>
+          Right(state.beginRollback())
+      }
+      afterChildren <- withClue(s"visiting children of $node") {
+        visitSubtrees(ksm)(nodes, children(node).toSeq, resolver, next)
+      }
+      exited = node match {
+        case _: Node.Rollback => afterChildren.endRollback()
+        case _ => afterChildren
+      }
+    } yield exited
+  }
+
+  // Fully visits the trees rooted at `roots` in execution order.
+  // For each subtree visited, additionally visit this subtree starting from the initial state
+  // and check that advancing the current state yields the same resulting state
+  private def visitSubtrees(ksm: ContractStateMachine[Unit])(
```

Comment:
Could you please add a ScalaDoc contract saying how the fields of the returned state are built from the transaction that has been traversed?